returnn/datasets/postprocessing.py

```diff
@@ -0,0 +1,253 @@
+        self._dataset.init_seq_order(epoch=epoch, seq_list=seq_list, seq_order=seq_order)
+        self._data_iter = enumerate(self._build_dataset_iter())
+        return True
+
+    def _collect_single_seq(self, seq_idx: int) -> Optional[DatasetSeq]:
+        while True:
+            try:
+                loaded_seq_idx, tensor_dict = next(self._data_iter)
+            except StopIteration:
+                return None
+            assert loaded_seq_idx <= seq_idx, "_collect_single_seq must be done monotonically"
+            if loaded_seq_idx != seq_idx:
+                continue
+            seq = DatasetSeq(features={k: t.raw_tensor for k, t in tensor_dict.data.items()}, seq_idx=seq_idx)
+            return seq
+
+    def _build_dataset_iter(self) -> Iterator[TensorDict]:
+        """
+        :return: an iterator applying both the segment level and across-segment transformations on the given dataset
+        """
+
+        def _iterate_dataset() -> Iterator[TensorDict]:
+            """
+            :return: generator providing data samples in the form of a TensorDict
+            """
+            data_keys = self._dataset.get_data_keys()
+
+            seq_index = 0
+            while self._dataset.is_less_than_num_seqs(seq_index):
+                self._dataset.load_seqs(seq_index, seq_index + 1)
+                data = {data_key: self._dataset.get_data(seq_index, data_key) for data_key in data_keys}
+                data["seq_tag"] = str_to_numpy_array(self._dataset.get_tag(seq_index))
+                tensor_dict = self._data_dict_to_tensor_dict(data)
+                if self._map_seq is not None:
+                    tensor_dict = self._map_seq(tensor_dict)
+                    assert isinstance(tensor_dict, TensorDict), "map_seq must produce a TensorDict"
+                yield tensor_dict
+                seq_index += 1
+
+        data_iter = _iterate_dataset()
+        if self._map_seq_stream is None:
+            return data_iter
+
+        data_iter = self._map_seq_stream(data_iter)
+        assert isinstance(data_iter, Iterator), "map_seq_stream must produce an Iterator"
+        return data_iter
+
+    def _data_dict_to_tensor_dict(self, data_dict: Dict[str, ndarray]) -> TensorDict:
+        """
+        :return: the given data dict converted to a TensorDict class
+        """
+
+        def _make_tensor(name: str, data: ndarray) -> Tensor:
+            sparse_dim = None
+            if data.dtype.name.startswith("str"):
+                dims = []
+                dtype = "string"
+            else:
+                dims, sparse_dim = self._dim_cache.get(name, (None, None))
+                dtype = data.dtype
+
+            if dims is None:
+                feature_dims = [
+                    Dim(dimension=v, name=f"{name}_dim{i + 1}")
+                    for i, v in enumerate(self._dataset.get_data_shape(name))
+                ]
+                dims = [Dim(dimension=None, name=f"{name}_frame"), *feature_dims]
```

Comment:
This here is not necessary. You should create the `Dim`s only once. You already did that before when you created the `TensorDict` object.